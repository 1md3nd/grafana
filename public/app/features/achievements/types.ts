export enum AchievementLevel {
  Novice = 1,
  Beginner = 2,
  Experienced = 3,
  Expert = 4,
  Wizard = 5,
}

export enum AchievementId {
  NavigateToDashboard = 'NavigateToDashboard', // DONE
  NavigateToExplore = 'NavigateToExplore', // DONE
  WatchIntroToGrafanaVideo = 'WatchIntroToGrafanaVideo', // TBD by achievements page
  ConnectYourFirstDatasource = 'ConnectYourFirstDatasource', // DONE
  UseExploreToMakeAQuery = 'UseExploreToMakeAQuery', // DONE
  AddExplorePanelToADashboard = 'AddExplorePanelToADashboard', // DONE
  AddATitleToAPanelInADashboard = 'AddATitleToAPanelInADashboard', // DONE
  AddADescriptionToAPanelInADashboard = 'AddADescriptionToAPanelInADashboard', // DONE
  ChangeTheTheme = 'ChangeTheTheme', // DONE
  ExploreKeyboardShortcuts = 'ExploreKeyboardShortcuts', // DONE
  ChangePanelSettings = 'ChangePanelSettings', // SKIP (not sure how to best implement this)
  ImplementDataLink = 'ImplementDataLink', // SKIP pain due to code being in grafana UI package
  AddTemplateVariable = 'AddTemplateVariable', // DONE
  AddDataTransformation = 'AddDataTransformation', // stopping point for today
  AddCanvasVisualization = 'AddCanvasVisualization',
  AddMetricValueElement = 'AddMetricValueElement',
  EnableCrosshairSharing = 'EnableCrosshairSharing',
  LegendChangeSeriesColor = 'LegendChangeSeriesColor',
  LegendShowSeries = 'LegendShowSeries',
  LegendHideSeries = 'LegendHideSeries',
  AddCustomThresholds = 'AddCustomThresholds',
  AddValueMapping = 'AddValueMapping',
  AddAdvancedDataLink = 'AddAdvancedDataLink',
  UseOuterJoinTransformation = 'UseOuterJoinTransformation',
  MakePublicDashboard = 'MakePublicDashboard',
  StreamDataToGrafana = 'StreamDataToGrafana',
}

export type Achievement = {
  id: AchievementId;
  title: string;
  description?: string;
  level: AchievementLevel;
  link?: string;
  video?: string;
  completed?: boolean;
};
