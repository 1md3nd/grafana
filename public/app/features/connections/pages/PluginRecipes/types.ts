export type PluginRecipe = {
  id: string;
  name: string;
  logo: string;
  summary: string;
  description?: string;
  steps: PluginRecipeStep[];
};

export type PluginRecipeStep<T = unknown> = {
  action: 'install-plugin' | 'display-info' | 'setup-dashboard' | 'prompt' | 'setup-alerts' | 'install-agent';
  name: string;
  description: string;
  settings: T;
  status: {
    code: StepStatus;
    message?: string;
  };
};

export enum StepStatus {
  Completed = 'Completed',
  NotCompleted = 'NotCompleted',
  Error = 'Error',
}

export type Screenshot = { name: string; url: string };

// Step - Install Plugin
// ---------------------
export type InstallPluginStepSettings = {
  plugin: {
    id: string;
    version: string;
  };
};

// Step - Prompt
// --------------
export type PromptStepSettings = {
  prompts: PluginRecipePrompt[];
};

export type PluginRecipePrompt = {
  label: string;
  description: string;
  type: string; // 'text', 'number', 'select', 'multiselect', 'radio'
  placeholder: string;
  defaultValue: string;
  options: Array<{
    name: string;
    value: string;
  }>;
};

// Step - Instruction
// -----------------
export type InstructionStepSettings = {
  instructionText: string; // The text of the instructions as Markdown
  instructionTestURL: string; // The URL to run the health-check against
  instructionTestExpectedHttpResponse: string; // The expected healthy HTTP response code (e.g. "200")
};

// Step - Dashboard
// -----------------
export type SetupDashboardStepSettings = {
  screenshots: Screenshot[];
};

// Step - Alerts
// -----------------
export type RecipeAlertRule = {
  namespace: string;
  group: string;
  name: string;
  summary: string;
};

export type SetupAlertsStepSettings = {
  alerts: RecipeAlertRule[];
};

// Step - Agent
// -----------------
export type RecipeMetric = {
  name: string;
  type?: string;
  description?: string;
};

export type InstallAgentStepSettings = {
  metrics: RecipeMetric[];
};

export function isSetupDashboardStep(step: PluginRecipeStep): step is PluginRecipeStep<SetupDashboardStepSettings> {
  return step.action === 'setup-dashboard';
}

export function isInstrucitonStep(step: PluginRecipeStep): step is PluginRecipeStep<InstructionStepSettings> {
  return step.action === 'display-info';
}

export function isPromptStep(step: PluginRecipeStep): step is PluginRecipeStep<PromptStepSettings> {
  return step.action === 'prompt';
}

export function isInstallPluginStep(step: PluginRecipeStep): step is PluginRecipeStep<InstallPluginStepSettings> {
  return step.action === 'install-plugin';
}

export function isSetupAlertsStep(step: PluginRecipeStep): step is PluginRecipeStep<SetupAlertsStepSettings> {
  return step.action === 'setup-alerts';
}

export function isInstallAgentStep(step: PluginRecipeStep): step is PluginRecipeStep<InstallAgentStepSettings> {
  return step.action === 'install-agent';
}
