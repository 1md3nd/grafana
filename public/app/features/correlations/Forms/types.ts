import { Correlation, CorrelationConfig } from '../types';

export interface FormDTO {
  sourceUID: string;
  targetUID: string;
  label: string;
  description: string;
  config: CorrelationConfig;
}

type FormDTOWithoutTarget = Omit<FormDTO, 'targetUID'>;
export type EditFormDTO = Partial<FormDTOWithoutTarget> &
  Pick<FormDTO, 'sourceUID' | 'config'> & { uid: Correlation['uid'] };
