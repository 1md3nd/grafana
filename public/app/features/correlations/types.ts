import { Correlation } from '@grafana/data';

export interface AddCorrelationResponse {
  correlation: Correlation;
}

export type GetCorrelationsResponse = Correlation[];

export interface CorrelationsApiResponse {
  message: string;
}

export interface CorrelationsErrorResponse extends CorrelationsApiResponse {
  error: string;
}

export interface CreateCorrelationResponse extends CorrelationsApiResponse {
  result: Correlation;
}

export interface UpdateCorrelationResponse extends CorrelationsApiResponse {
  result: Correlation;
}

export interface RemoveCorrelationResponse {
  message: string;
}

export type GetCorrelationsParams = {
  page: number;
};

export type RemoveCorrelationParams = Pick<Correlation, 'sourceUID' | 'uid'>;
export type CreateCorrelationParams = Omit<Correlation, 'uid'>;
export type UpdateCorrelationParams = Omit<Correlation, 'targetUID'>;
