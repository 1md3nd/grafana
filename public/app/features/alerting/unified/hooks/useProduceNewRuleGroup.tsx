import { Action } from '@reduxjs/toolkit';
import { useCallback } from 'react';

import { dispatch, getState } from 'app/store/store';
import { RuleGroupIdentifier } from 'app/types/unified-alerting';
import { RulerGrafanaRuleDTO, RulerRuleDTO } from 'app/types/unified-alerting-dto';

import { alertRuleApi } from '../api/alertRuleApi';
import { deleteRuleAction, pauseRuleAction, ruleGroupReducer } from '../reducers/ruler/ruleGroups';
import {
  fetchPromAndRulerRulesAction,
  fetchRulesSourceBuildInfoAction,
  getDataSourceRulerConfig,
} from '../state/actions';

import { mergeRequestStates } from './mergeRequestStates';

type ProduceNewRuleGroupOptions = {
  /**
   * Should we dispatch additional actions to ensure that other (non-RTKQ) caches are cleared?
   */
  refetchAllRules?: boolean;
};

function useProduceNewRuleGroup() {
  const [fetchRuleGroup, fetchRuleGroupState] = alertRuleApi.endpoints.getRuleGroupForNamespace.useLazyQuery();
  const [deleteRuleGroup, deleteRuleGroupState] = alertRuleApi.endpoints.deleteRuleGroupFromNamespace.useLazyQuery();
  const [updateRuleGroup, updateRuleGroupState] = alertRuleApi.endpoints.updateRuleGroupForNamespace.useMutation();

  const produceNewRuleGroup = async (
    ruleGroup: RuleGroupIdentifier,
    action: Action,
    options?: ProduceNewRuleGroupOptions
  ) => {
    const { dataSourceName, groupName, namespaceName } = ruleGroup;

    // @TODO we should really not work with the redux state (getState) here
    await dispatch(fetchRulesSourceBuildInfoAction({ rulesSourceName: dataSourceName }));
    const rulerConfig = getDataSourceRulerConfig(getState, dataSourceName);

    const currentRuleGroup = await fetchRuleGroup({
      rulerConfig,
      namespace: namespaceName,
      group: groupName,
    }).unwrap();

    // @TODO convert rule group to postable rule group – TypeScript is not complaining here because
    // the interfaces are compatible but it _should_ complain
    const newRuleGroup = ruleGroupReducer(currentRuleGroup, action);

    const deleteEntireGroup = newRuleGroup.rules.length === 0;

    // if we have no more rules in the group, we delete the entire group, otherwise just update the rule group
    const updateOrDelete = () => {
      if (deleteEntireGroup) {
        return [
          deleteRuleGroup({
            rulerConfig,
            namespace: namespaceName,
            group: groupName,
          }).unwrap(),
          mergeRequestStates(fetchRuleGroupState, deleteRuleGroupState),
        ] as const;
      }

      return [
        updateRuleGroup({
          rulerConfig,
          namespace: namespaceName,
          payload: newRuleGroup,
        }).unwrap(),
        mergeRequestStates(fetchRuleGroupState, updateRuleGroupState),
      ] as const;
    };

    const [updateOrDeleteResult, state] = updateOrDelete();

    if (options?.refetchAllRules) {
      // refetch rules for this rules source
      // @TODO remove this when we moved everything to RTKQ – then the endpoint will simply invalidate the tags
      dispatch(fetchPromAndRulerRulesAction({ rulesSourceName: ruleGroup.dataSourceName }));
    }

    return [updateOrDeleteResult, state];
  };

  // @TODO merge loading state with the fetching state
  return [produceNewRuleGroup, updateRuleGroupState] as const;
}

export function usePauseRuleInGroup() {
  const [produceNewRuleGroup, produceNewRuleGroupState] = useProduceNewRuleGroup();

  const pauseFn = useCallback(
    async (ruleGroup: RuleGroupIdentifier, rule: RulerGrafanaRuleDTO, pause: boolean) => {
      const uid = rule.grafana_alert.uid;
      const action = pauseRuleAction({ uid, pause });

      return produceNewRuleGroup(ruleGroup, action);
    },
    [produceNewRuleGroup]
  );

  return [pauseFn, produceNewRuleGroupState] as const;
}

export function useDeleteRuleFromGroup() {
  const [produceNewRuleGroup, produceNewRuleGroupState] = useProduceNewRuleGroup();

  const deleteFn = useCallback(
    async (ruleGroup: RuleGroupIdentifier, rule: RulerRuleDTO) => {
      const action = deleteRuleAction({ rule });

      return produceNewRuleGroup(ruleGroup, action, { refetchAllRules: true });
    },
    [produceNewRuleGroup]
  );

  return [deleteFn, produceNewRuleGroupState] as const;
}
