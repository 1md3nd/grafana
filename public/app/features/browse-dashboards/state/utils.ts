import { DashboardViewItem } from 'app/features/search/types';

import { BrowseDashboardsState } from '../types';

export function findItem(
  childrenCollection: BrowseDashboardsState['children'],
  uid: string
): DashboardViewItem | undefined {
  for (const key in childrenCollection) {
    const children = childrenCollection[key];

    if (!children) {
      continue;
    }

    for (const child of children.items) {
      if (child.uid === uid) {
        return child;
      }
    }
  }

  return undefined;
}

export function getPaginationPlaceholders(amount: number, parentUID: string | undefined, level: number) {
  return new Array(amount).fill(null).map((_, index) => {
    return {
      parentUID,
      level,
      isOpen: false,
      item: {
        kind: 'ui' as const,
        uiKind: 'pagination-placeholder' as const,
        uid: `${parentUID}-pagination-${index}`,
      },
    };
  });
}

type KeyableArgs = {
  parentUID: string | undefined;
  excludeKinds?: string[];
};

export function getChildrenStateKey({ parentUID, excludeKinds }: KeyableArgs) {
  return JSON.stringify({
    parentUID: parentUID ?? '$$special_uid_for_grafana_root_folder', // JOSH TODO: we probably don't need to set a root folder uid, and just rely on it being undefined
    excludeKinds: excludeKinds ?? [],
  });
}
