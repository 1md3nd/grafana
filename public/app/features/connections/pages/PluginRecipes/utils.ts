import { applyStep } from './api';
import { PluginRecipeAction, PluginRecipe } from './types';

// Finds the steps that can be auto-applied starting from a certain index
// (It is used to automatically apply steps that don't need a user action, but stop when a user action is needed)
export const getAutoApplicapleStepIndexes = (recipe?: PluginRecipe, fromStepIndex = 0) => {
  if (!recipe) {
    return [];
  }

  const stepIds = [];

  for (const [i, step] of recipe.steps.slice(fromStepIndex).entries()) {
    if (step.action === PluginRecipeAction.DisplayInfo || step.action === PluginRecipeAction.Prompt) {
      break;
    }

    stepIds.push(i + fromStepIndex);
  }
  return stepIds;
};

// Can be used to either start or continue an install process
export const installRecipe = async (recipe?: PluginRecipe, fromStepIndex = 0) => {
  if (!recipe) {
    return;
  }

  const autoApplicableStepIndexes = getAutoApplicapleStepIndexes(recipe, fromStepIndex);
  await Promise.all(autoApplicableStepIndexes.map((index) => applyStep(recipe.id, index)));
};
