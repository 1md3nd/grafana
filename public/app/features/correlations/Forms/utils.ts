import { Correlation } from '@grafana/data';

type CorrelationBaseData = Pick<Correlation, 'uid' | 'sourceUID' | 'targetUID'>;

export const getInputId = (inputName: string, correlation?: CorrelationBaseData) => {
  if (!correlation) {
    return inputName;
  }

  return `${inputName}_${correlation.sourceUID}-${correlation.uid}`;
};
