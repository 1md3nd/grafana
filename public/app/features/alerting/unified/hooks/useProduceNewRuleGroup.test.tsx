import { UseQueryStateResult } from '@reduxjs/toolkit/dist/query/react/buildHooks';
import { waitFor, renderHook } from 'test/test-utils';

import { setBackendSrv } from '@grafana/runtime';
import { backendSrv } from 'app/core/services/backend_srv';
import { RuleGroupIdentifier } from 'app/types/unified-alerting';

import { setupMswServer } from '../mockApi';
import { grafanaRulerGroupName, grafanaRulerNamespace, grafanaRulerRule } from '../mocks/alertRuleApi';
import { GRAFANA_RULES_SOURCE_NAME } from '../utils/datasource';

import { usePauseRuleInGroup } from './useProduceNewRuleGroup';

setupMswServer();

beforeAll(() => {
  setBackendSrv(backendSrv);
});

it('should be able to pause a rule', async () => {
  const ruleGroupIdentifier: RuleGroupIdentifier = {
    namespace: grafanaRulerNamespace.uid,
    group: grafanaRulerGroupName,
    ruleSourceName: GRAFANA_RULES_SOURCE_NAME,
  };

  expect(grafanaRulerRule).toHaveProperty('grafana_alert.is_paused', false);

  const { result } = renderHook(() => {
    const [pauseRule, pauseState] = usePauseRuleInGroup();
    pauseRule(ruleGroupIdentifier, grafanaRulerRule, true);

    return pauseState;
  });

  await waitFor(() => {
    expect(result.current).toHaveProperty('status', 'fulfilled');
  });

  // a bit hacky to inspect what was passed in to RTKQ â€“ open to ideas
  expect(result.current).toHaveProperty('originalArgs.payload.rules[0].grafana_alert.is_paused', true);

  // @ts-ignore
  expect(result.current).toMatchSnapshot({
    fulfilledTimeStamp: expect.any(Number),
    startedTimeStamp: expect.any(Number),
    requestId: expect.any(String),
  });
});
