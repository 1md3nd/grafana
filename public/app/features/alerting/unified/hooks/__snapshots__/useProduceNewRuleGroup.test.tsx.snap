// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to pause a rule 1`] = `
{
  "data": {
    "message": "updated",
    "updated": [],
  },
  "endpointName": "updateRuleGroupForNamespace",
  "fulfilledTimeStamp": Any<Number>,
  "isError": false,
  "isLoading": false,
  "isSuccess": true,
  "isUninitialized": false,
  "originalArgs": {
    "namespace": "uuid020c61ef",
    "payload": {
      "interval": "1m",
      "name": "grafana-group-1",
      "rules": [
        {
          "annotations": {
            "summary": "Test alert",
          },
          "for": "5m",
          "grafana_alert": {
            "condition": "A",
            "data": [
              {
                "datasourceUid": "datasource-uid",
                "model": {
                  "datasource": {
                    "type": "prometheus",
                    "uid": "datasource-uid",
                  },
                  "expression": "vector(1)",
                  "queryType": "alerting",
                  "refId": "A",
                },
                "queryType": "alerting",
                "refId": "A",
                "relativeTimeRange": {
                  "from": 1000,
                  "to": 2000,
                },
              },
            ],
            "exec_err_state": "Error",
            "is_paused": true,
            "namespace_uid": "uuid020c61ef",
            "no_data_state": "NoData",
            "rule_group": "grafana-group-1",
            "title": "Grafana-rule",
            "uid": "4d7125fee983",
          },
          "labels": {
            "region": "nasa",
            "severity": "critical",
          },
        },
      ],
    },
    "rulerConfig": {
      "apiVersion": "legacy",
      "dataSourceName": "grafana",
    },
  },
  "requestId": Any<String>,
  "reset": [Function],
  "startedTimeStamp": Any<Number>,
  "status": "fulfilled",
}
`;
